File: CloudFormation/src/main/java/com/intellij/aws/cloudformation/references/CloudFormationReferenceBase.java
package com.intellij.aws.cloudformation.references;

import com.intellij.json.psi.JsonElementGenerator;
import com.intellij.json.psi.JsonLiteral;
import com.intellij.json.psi.JsonStringLiteral;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReferenceBase;
import com.intellij.util.IncorrectOperationException;
import org.jetbrains.annotations.NotNull;

public abstract class CloudFormationReferenceBase extends PsiReferenceBase<JsonLiteral> {
  public CloudFormationReferenceBase(@NotNull JsonLiteral element) {
    super(element);
  }

  @NotNull
  @Override
  public Object[] getVariants() {
    return getCompletionVariants();
  }

  @Override
  public PsiElement handleElementRename(String newElementName) throws IncorrectOperationException {
    final JsonStringLiteral newElement = new JsonElementGenerator(myElement.getProject()).createStringLiteral(newElementName);
    return myElement.replace(newElement);
  }

  @NotNull
  public abstract String[] getCompletionVariants();
}

File: CloudFormation/src/main/kotlin/com/intellij/aws/cloudformation/references/CloudFormationReferenceBase.kt
package com.intellij.aws.cloudformation.references

import com.intellij.aws.cloudformation.CloudFormationParser
import com.intellij.aws.cloudformation.model.CfnScalarValueNode
import com.intellij.aws.cloudformation.ofType
import com.intellij.json.psi.JsonElementGenerator
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReferenceBase

abstract class CloudFormationReferenceBase(element: PsiElement) : PsiReferenceBase<PsiElement>(element) {
  val scalarNode: CfnScalarValueNode

  init {
    val parsed = CloudFormationParser.parse(element.containingFile)
    scalarNode = parsed.getCfnNodes(element).ofType<CfnScalarValueNode>().single()
  }

  override fun getVariants(): Array<Any> {
    return getCompletionVariants().toTypedArray()
  }

  override fun handleElementRename(newElementName: String): PsiElement {
    val newElement = JsonElementGenerator(myElement.project).createStringLiteral(newElementName)
    return myElement.replace(newElement)
  }

  abstract fun getCompletionVariants(): List<String>
}
